### ISSUE TITLE
Option to Log In Without Linking to Twitter, GitHub, or Google account

### ISSUE THREAD

**Issue by eyesmo** [vsoch/freegenes#46]
Hi! It would be good to give users the option to create a login that is unique to the FreeGenes website, since requiring a link to and account on one of the major social media sites or a GitHub account might restrict some people's ability to create logins, or might make some people uncomfortable/uncertain about what information from those accounts FreeGenes can access. Age of surveillance capitalism and all that.

**Comment by Koeng101** [comment]
@vsoch how hard is it to run a small application that is a BioBricks specific login? (I'm guessing there is an open source project that implements social login)

I think it would be a good option for people (I would even prefer to login using a BioBricks specific login than using Google or Github)

**Comment by vsoch** [comment]
hey @eyesmo thanks for your comments! We had a discussion briefly about this in discord, and I don't think it's a good idea. It's generally not good practice to store actual credentials on a server - it's just not secure. With OAuth2 we use tokens and refresh tokens, and both the user and admin can at any time revoke or refresh tokens. An important note to stress to your users is that we don't get any sort of permission to their accounts with OAuth2, it's just asking for read only of the email address and username, which is very reasonable (and much better practice than needing an email plus password / username). 

I would definitely be open to considering other kinds of institution auth, or OAuth 2, but using the bare bones Django username /password is not something I'm wanting to do, especially since from a technical perspective it doesn't add anything, but actually can make the server design less optimal/secure. Please take a look at plugins that can support other auth types -> https://vsoch.github.io/freegenes/docs/plugins/ along with the huge list that is supported by social auth (that I could implement). https://python-social-auth.readthedocs.io/en/latest/backends/. And of course let me know if there is another institutional level auth that you'd like us to look at.

**Comment by vsoch** [comment]
BioBricks would need to implement OAuth2, then it would work!

**Comment by vsoch** [comment]
What is the backend for biobricks? Also I just noticed this towards the bottom - looks suspicious?

![image](https://user-images.githubusercontent.com/814322/64718961-438c6200-d495-11e9-88f7-5c201ab12169.png)

The link goes to "bestgrammarchecker.net" which you can bet I'm not going to click, haha. :)

**Comment by vsoch** [comment]
Oh I see it's a wordpress site.

**Comment by eyesmo** [comment]
@vsoch If it's only getting read-only access to the email address and password, that should be fine. But in that case it might be good to find a way to get Twitter's third party login disclaimer to reflect that, since at the moment it says that FreeGenes gains access to a lot more info and control of the Twitter account than just the text of the email address and password.
<img width="696" alt="Screen Shot 2019-09-11 at 12 08 39 PM" src="https://user-images.githubusercontent.com/2289040/64719169-4daa6280-d48d-11e9-9ccf-4edfffb2753f.png">

**Comment by Koeng101** [comment]
If someone logins using a raw email + password, I think that is an acknowledgement of the security vulnerabilities in exchange for independence from large companies. 

I don't think we're really asking on a technical side that login be implemented within the node itself (though that would be preferred, but I understand why you could be opposed). Rather, looking at **running** a different application (separate code base, separate service) that would act in the place of any of those login systems. 

A quick browse online shows there are open source applications that do this. Could we just deploy and maintain a deployment of one of those?

---

Yea the BioBricks website is just a wordpress site, haven't really interacted with it in a very long while.

**Comment by vsoch** [comment]
@eyesmo to be clear, OAuth2 does not share any kind of password from Twitter - a token and refresh token are issues (by Twitter) that can be revoked by the calling application or user at any time. There are no passwords shared. This is much more secure than having the user store a password on the server. If the server is compromised, the tokens can be revoked, no harm done. If the password is on the server, well you're toast.

**Comment by eyesmo** [comment]
^So what's the deal with all the info and control that Twitter says FreeGenes is being granted through the 3rd party login?

**Comment by vsoch** [comment]
Fixed

![image](https://user-images.githubusercontent.com/814322/64719546-77b45280-d496-11e9-9028-64780c9aff06.png)

**Comment by vsoch** [comment]
Just a tab I missed when I was setting it up :) Now it's read only access to whatever is already public. Thanks for catching this! Twitter is different from the other OAuth2 because you set the permissions via their interface (and not as variables in your own code).

**Comment by Koeng101** [comment]
Don't you normally store the password as a hash on the server anyway? I guess I can't see why you are toast if the server has your password hash.

**Comment by vsoch** [comment]
Yes that's what Django would do if we implemented it.

**Comment by Koeng101** [comment]
But if your password is hashed, then why are you toast if they take over server?

**Comment by vsoch** [comment]
Because the secret is on the server.

**Comment by vsoch** [comment]
:/

**Comment by vsoch** [comment]
This is trying to engineer a hacky solution for what is a solved problem.

**Comment by Koeng101** [comment]
Looks like using sha256 IS the solution for this problem. Django uses it with the PBKDF2, which shouldn't be breakable with a server-side secret key https://docs.djangoproject.com/en/2.2/topics/auth/passwords/

This kind of one way hashing is also used on Unix systems https://en.wikipedia.org/wiki/Passwd (sudo vim /etc/shadow to see what type of encryption your system is using, mine is using sha512) which has been in use since about 1992. 

What Django password method were you thinking of? AFAIK their password management documentation doesn't say anything about private key signing.

**Comment by vsoch** [comment]
https://docs.djangoproject.com/en/2.2/topics/auth/passwords/

**Comment by Koeng101** [comment]
Ok - I'm guessing that the reason for the wontfix label is that Django does do it server side, so if they get into the server, they can simply intercept the password coming in? To be completely honest, I still don't fully understand the aversion to using Django login after reading their documentation.

Regardless, what about running a separate login service as posted above? If someone uses it, they are willing consenting to using our system, and so I don't find any ethical reasons of why that is bad (but can think of a few with current social login)

**Comment by vsoch** [comment]
I'll ask my colleagues - for Stanford apps / servers you would never see this sort of thing.

**Comment by akkornel** [comment]
Hello!  

I'm one of @vsoch's coworkers, and was chatting with her about this earlier.

I definitely see valid points on both sides:

(In the below, I use "social login site" to refer to sites like Twitter, Facebook, and also GitHub, etc.)

• I understand that someone might not have an account at a social login site, or might choose to not use that site to authenticate to other sites.  Using a social login site to log in does tell that site that you are a user of this site; you have to trust that the social login site would not use this new information, and that they won't hold on to that information for longer than is absolutely necessary.  In addition, there are people who can't use social login sites.  For example, I understand that GitHub locks (or otherwise limits) accounts of people in certain countries (as per https://help.github.com/en/articles/github-and-trade-controls).  And I understand that most sites do not (officially, at least) want you to create an account if you are under 13.  Or the site(s) may be blocked by something on 'your end' of the connection (for example, a home, ISP, or country-wide firewall).

• At Stanford, we are trained to avoid static password as much as possible.  For example, see Stanford's [Minimum Security Standards for Applications](https://uit.stanford.edu/guide/securitystandards#security-standards-applications), which directs application developers to use [Stanford Login](https://uit.stanford.edu/service/saml/).  Storing a passwords means you have to handle the password in cleartext form (even if briefly), which brings the risk of leaking (for example, into log files or debug messages).  A leaked list of hashed/derived password could, for example, be used to test against an existing password list; if an attacker has a password that a user has used before, it would be fairly quick to check against a hashed/derived password, which increases the chances that the password is still in use elsewhere.  And in general, it increases the amount of stress on the part of the people developing this code.

(And to be clear, in everything I am saying, I am talking about credentials used by humans.  I am not talking about things like API keys, which is a different discussion.)

Given all of those points, I still cannot recommend that @vsoch add support for local credentials.  But instead, I am suggesting that she add support for [Globus Auth](https://docs.globus.org/api/auth/), as described at https://globus-integration-examples.readthedocs.io/en/latest/django.html.

Globus Auth has a number of benefits that I think would help this use case:

• It has already been used in Django apps before, so adding support for it should (hopefully!) not be difficult.
• Globus Auth is run by Globus, which is part of the University of Chicago.  So, instead of trusting in a US-based publicly-traded private company, you are trusting in a US-based not-for-profit (specifically, a [501(c)(3)](https://projects.propublica.org/nonprofits/organizations/362177139)) private university.  If your identity provider is authenticated via CILogon, which is run by the University of Illinois at Urbana-Champaign, so you may also be trusting in a US-based public university.
• Globus Auth supports logins from a large list of institutional identity providers from across the world (thanks to the aforementioned relationship with CILogon).  If this app is meant to be used by individual labs, then adding Globus Auth makes it possible to log in with your own institution's login.  That is beneficial in that the user doesn't have to remember a separate username/password (or reuse an existing password), and the lab manager doesn't have to remember to deactivate a person's account when they leave the lab.
• For labs that want more control over who can log in, or who are required to do so by their institution, Globus Auth allows restricting logins to a specific identity provider.
• For users who still want to have a dedicated username and password, not tied to anything else, you can create a [Globus ID](https://www.globusid.org), which you use exclusively for this purpose.

(Full disclosure: I am one of the named contacts on Stanford's Globus subscription, so I do have a bit of a bias in favor of Globus Auth.  I have interacted with Globus staffers before.  I am not getting any compensation — from them or anyone else — for this recommendation.)

It's worth noting, I was considering suggesting the use of [CILogon](https://cilogon.org) instead of Globus Auth, because Globus Auth actually 'outsources' most authentications to CILogon.  But there are a few reasons why I did not make this suggestion:

• Globus Auth supports a few third-party identity providers that CILogon does not, I think mainly Globus subscribers who are not in CILogon's list.
• Globus Auth supports Globus ID, for people who do not want an account tied to anything else.  With CILogon, as far as I know, there is no such option.

I'm sorry I do not agree with this request, but I hope the suggestion of Globus Auth will help to address the issue that was raised here!

**Comment by vsoch** [comment]
Thanks @akkornel!  :D 

@Koeng101 and @eyesmo I brought in Karl because he is a wealth of information, both for policy and implementation details (well, this is probably very obvious! :) ). I'd definitely be happy to figure out Globus authentication, it's a method that I strongly support too.

**Comment by eyesmo** [comment]
The Globus Auth option sounds great! Especially paired with a bit of text letting people know that this is the option for people not able to/not comfortable with logging in through the other options, and including a link explaining what a Globus ID is and how to set one up. Thanks @vsoch and @akkornel!
